src/Middleware/HttpLogging/src/ResponseBufferingStream.cs

```diff
@@ -0,0 +1,177 @@
+{
+    /// <summary>
+    /// Stream that buffers reads 
+    /// </summary>
+    internal sealed class ResponseBufferingStream : BufferingStream, IHttpResponseBodyFeature
+    {
+        private readonly IHttpResponseBodyFeature _innerBodyFeature;
+        private readonly int _limit;
+        private PipeWriter? _pipeAdapter;
+
+        private readonly HttpContext _context;
+        private readonly List<MediaTypeState> _encodings;
+        private readonly HttpLoggingOptions _options;
+        private Encoding? _encoding;
+
+        private static readonly StreamPipeWriterOptions _pipeWriterOptions = new StreamPipeWriterOptions(leaveOpen: true);
+
+        internal ResponseBufferingStream(IHttpResponseBodyFeature innerBodyFeature,
+            int limit,
+            ILogger logger,
+            HttpContext context,
+            List<MediaTypeState> encodings,
+            HttpLoggingOptions options)
+            : base(innerBodyFeature.Stream, logger)
+        {
+            _innerBodyFeature = innerBodyFeature;
+            _innerStream = innerBodyFeature.Stream;
+            _limit = limit;
+            _context = context;
+            _encodings = encodings;
+            _options = options;
+        }
+
+        public bool FirstWrite { get; private set; }
+
+        public Stream Stream => this;
+
+        public PipeWriter Writer
+        {
+            get
+            {
+                if (_pipeAdapter == null)
+                {
+                    _pipeAdapter = PipeWriter.Create(Stream, _pipeWriterOptions);
+                }
+
+                return _pipeAdapter;
+            }
+        }
+
+        public Encoding? Encoding { get => _encoding; }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            Write(buffer.AsSpan(offset, count));
+        }
+
+        public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)
+        {
+            return TaskToApm.Begin(WriteAsync(buffer, offset, count), callback, state);
+        }
+
+        public override void EndWrite(IAsyncResult asyncResult)
+        {
+            TaskToApm.End(asyncResult);
+        }
+
+        public override void Write(ReadOnlySpan<byte> span)
+        {
+            var remaining = _limit - _bytesBuffered;
+            var innerCount = Math.Min(remaining, span.Length);
+
+            OnFirstWrite();
+
+            if (span.Slice(0, innerCount).TryCopyTo(_tailMemory.Span))
+            {
+                _tailBytesBuffered += innerCount;
+                _bytesBuffered += innerCount;
+                _tailMemory = _tailMemory.Slice(innerCount);
+            }
+            else
+            {
+                BuffersExtensions.Write(this, span.Slice(0, innerCount));
+            }
+
+            _innerStream.Write(span);
+        }
+
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
```

Comment:
Did you forget to override the `ReadOnlyMemory<byte>` overload.

Reply:
Yes, but doesn't the default on still call WriteAsync(byte[], offset, count) https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L615

Reply:
You need override it and delegate this call to the `ReadOnlyMemory<byte>` overload